[MIG][9.0] sale automatic workflow

SodexisTeam opened the pull request:
https://odoo-community.org/groups/contributors-15/contributors-39175

sebastienbeau committed:
[ADD] add sale_automatic_workflow, a module to automatized the sale order workflow

sebastienbeau committed:
[IMP] sale_automatic_workflow : add default value for workflow method

Commit:
[FIX] sale_automatic_workflow : fix invoice creation from picking

Commit:
[IMP] sale_automatic_workflow: redesign of sale order view

sebastienbeau committed:
[FIX] sale_automatic_workflow : add missing security rule

sebastienbeau committed:
[FIX] sale_automatic_workflow : reconciled should be done by the cron task, indeed if we did it in the workflow we can not anymore unreconcile an invoice because it will be re-reconcile directly

sebastienbeau committed:
[FIX] sale_automatic_wokflow : fix the option force invoice date, indeed this feature was still not reimplemented

sebastienbeau committed:
[REF] refactor/clean code due to orm change in 6.1

sebastienbeau committed:
[MERGE] sale_automatic_workflow : merge fix

sebastienbeau committed:
[IMP] sale_automatic_workflow : use context manager for intermediary commit

sebastienbeau committed:
[IMP] sale_automatic_workflow : add doc string for commit_now and set to False params raise_error as it's a boolean

Commit:
Ir.cron protected by a noupdate flag.

sebastienbeau committed:
[IMP] sale_automatic_workflow : invoice validation and reconciliation should be process individualy

sebastienbeau committed:
[REF] import the commit_now from framework_helpers as it have been extracted

sebastienbeau committed:
[FIX] Open a new cursor before commiting it, thanks to Guewen ideas

sebastienbeau committed:
[IMP] set correctly the version number 6.1.X

sebastienbeau committed:
[IMP] sale_automatic_workflow: improve invoice reconciliation by supporting exchange write off

sebastienbeau committed:
[MIGR] sale_automatic_workflow: move code in the correct file

sebastienbeau committed:
[MIGR] sale_automatic_workflow: migration module to V7

Commit:
[FIX] sale_automatic_workflow: replace dependencies on sale_quick_payment by sale_quick_payment

Commit:
[IMP] sale_automatic_worflow: reconcile with the move lines instead of the voucher

Commit:
[IMP] sale_automatic_workflow: review

Commit:
[FIX] since openerp 7, the crons callee have their own cr, so we don't need anymore to open a new one

Commit:
[IMP] review the workflow job, get rid of the framework helpers dependency

Commit:
[FIX] remove the field 'validate_manufactoring_order', not implemented actually

Commit:
[IMP] use an onchange to apply the default values (connectors will have to use this onchange

Commit:
[FIX] rename onchange_workflow_process to onchange_workflow_process_id

Commit:
[IMP] sale_automatic_workflow: use the same words than openerp for order_policy, remove the 'postpaid' removed in openerp

Commit:
[IMP] sale_automatic_workflow: display the workflow below the payment method

Commit:
[IMP] sale_automatic_workflow: the onchange apply the workflow

Commit:
[FIX] forbid the unlink of a workflow used by a sale order

Commit:
[REF] sale_automatic_workflow: remove the field 'days_before_cancel', moved in connector_ecommerce

Commit:
[CHG] sale_automatic_workflow: 'validate_order' is now a boolean, because the orders are now created or not based on the payment status

Commit:
[FIX] noupdate on the automatic workflow

Commit:
[FIX] the workflow_process_id is not transferred to picking, due to a big and incomprehensible jiggery-pokery between stock.picking and stock.picking.out.

The field have to be created on stock.picking so the sale order is able to write the field on creation (in sale._create_pickings_and_procurements), but thereby we can't search a stock.picking.out by this field.

Commit:
[ADD] sale_automatic_workflow: automatic validation of sales orders according to the workflow

Commit:
[IMP] better field descriptions (respect the sale order ones) and change the view inheritance due to changes in sale_payment_method

Commit:
[FIX] _prepare_invoice is called on stock.picking and not on stock.picking.out

Commit:
[IMP] rename Sale Workflow Method to Automatic Workflow

Commit:
[IMP] the option to create an invoice now allows to choose if it is created on the confirmation of the sale order or after the delivery

Commit:
[FIX] updated base data according to last change and addded a manual workflow

Commit:
[FIX] do not activate force invoice date in the base workflows

Commit:
[ADD] sale_automatic_workflow: french translation

sebastienbeau committed:
[FIX] when a sale order is in exception we should stop to try to validate it every minutes, performance killer. So the best is to only try to validate the sale order that don't have any exception

Commit:
[FIX] use explicit relative import, use standard orm.Model

Commit:
[FIX] the exception variable (e) is not used

Commit:
[ADD] authorship

Commit:
[FIX] unused import

Commit:
[CHG] use isinstance rather than hasattr

Commit:
[IMP] docstring of sale.workflow.process

Commit:
[FIX] fr_FR.po files not loaded, should be named fr.po

sebastienbeau committed:
[REF] adapte sale_automatic_workflow due to the change as sale order are linked to a account move line

sebastienbeau committed:
[IMP] sle_automatic_workflow : add partial reconciliation

Commit:
[ADD] Camptocamp as author alongside Akretion

sebastienbeau committed:
[FIX] remove dependency on sale_exception and give the posibility to inherit easily the domain for the sale order validation

Commit:
[MRG] from master

sebastienbeau committed:
[FIX] remove dependency on sale_exception and give the posibility to inherit easily the domain for the sale order validation

Commit:
[FIX] missing argument in _get_sum_... methods, sale_order.payment_ids is already move lines, so we should not get their 'line_id'

Commit:
use method to get the move lines

albertofl committed:
Tranlations es for modules

Commit:
set the flag 'installable' to False on the modules

Commit:
move modules to migrate in __unported__

Commit:
Fix flake8 errors

Commit:
Shorten long lines with > 79 chars

Commit:
Remaining long lines in sale_automatic_workflow/__init__.py

rdeheele committed:
[ADD] display a warning related to a wkfl process

ex: if the user passes a sale order from the manual to the automatic workflow
process, it can be useful to inform him that a cron could automatically confirm
the order in a close future.
it warns the user via an onchange pop-up box.

rdeheele committed:
[UPD] updates help on warning field

Commit:
String statement has no effect: move the docstring above the imports

Commit:
Merge 7.0 into 8.0

Conflicts:
	.travis.yml
	README.md
	__unported__/product_links/__openerp__.py
	__unported__/sale_automatic_workflow/stock.py

Commit:
Move sale_automatic_workflow from __unported__ to root folder

Commit:
Set sale_automatic_workflow to installable

Commit:
Extract module's description in README.rst

Commit:
Remove dependency on sale_payment_method

I want to remove this dependency so we can use an automatic workflow
without using the payment methods.

A third module will surely be necessary.

Commit:
Remove changes to the model payment.method

Commit:
Remove what depends on sale_payment_method

It will be reimplemented later in a link module between
sale_payment_method and sale_automatic_workflow

Commit:
Started to implement a test

Commit:
stock.picking.out has passed away and methods have changed

Commit:
Add a page for the Automation information

Commit:
Deprecated comment

Commit:
Test fails on the last assert

Because workflow_process_id is not propagated to the stock.picking

Commit:
Check if invoice is generated and opened

Commit:
Propagate the workflow to the picking

Test pass now

Commit:
New API for sale.workflow.process

Commit:
New api account.invoice

Commit:
New API for sale.order

Commit:
New API for stock.picking

Commit:
Rename stock.py to stock_picking.py

Commit:
remove dead xml

Commit:
New api for automatic.workflow.job

Commit:
Add tests for propagation of date and workflow

Commit:
Put fields in groups so the labels are displayed

Commit:
Updated pot and fr.po

Commit:
Unused import

Commit:
Typo, added an uppercase

Commit:
Do not propagate sale sale automatic workflow when several orders are grouped in one picking

If we have several sales orders in one procurement group,
we group them in one picking. The possibilities when we have
several orders for one picking are:
1. take the workflow of the first sale order
2. do not propagate the workflow at all
3. propagate the workflow if all the workflows are the same
The solution 1. is bad, the 3. could be hard to understand and
unpredictable. The solution 2. takes no assumption on what is
expected.

gurneyalex committed:
Add OCA as author of OCA addons

In order to get visibility on https://www.odoo.com/apps the OCA board has
decided to add the OCA as author of all the addons maintained as part of the
association.

bguillot committed:
[FIX] change called method to confirm sale orders

pedrobaeza committed:
Merge pull request #32 from guewen/8.0-sale_automatic_worflow-many-so

Do not propagate sale sale automatic workflow when several orders are grouped in one picking

yvaucher committed:
Add missing default oca icons

damdam-s committed:
[IMP] add relation to sales team

damdam-s committed:
[IMP] add group base.group_multi_salesteams on 'section_id' fields in views

sbidoul committed:
[UPD] prefix versions with 8.0

oca-transbot committed:
OCA Transbot updated translations from Transifex

Commit:
sale_automatic_workflow : add default journal on workflow process

oca-transbot committed:
OCA Transbot updated translations from Transifex

Commit:
sale_automatic_workflow : create invoice for sale with only product service

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
Merge pull request #254 from acsone/8.0-sale_automatic_workflow_create_invoice-lga

sale_automatic_workflow : create invoice for sale with only service

oca-transbot committed:
OCA Transbot updated translations from Transifex

atchuthan committed:
Merge branch '8.0-extract' into 9.0-sale_automatic_workflow

guewen commented on sale_automatic_workflow/README.rst:
```diff
@@ -0,0 +1,23 @@
+  * Confirm the picking
+
+This module is used by Magentoerpconnect and Prestashoperpconnect.
+It is well suited for other E-Commerce connectors as well.
```
Could you apply the OCA's template?  https://raw.githubusercontent.com/OCA/maintainer-tools/master/template/module/README.rst

guewen commented on sale_automatic_workflow/automatic_workflow_job.py:
```diff
@@ -0,0 +1,122 @@
+steps of the workflow in the same transaction.
+
+If my explanation is not clear, contact me by email and I will improve
+it: sebastien.beau@akretion.com
```
I think this explanation is irrelevant now that sales workflows are gone and can be removed.

lmignon commented on sale_automatic_workflow/__openerp__.py:
```diff
@@ -0,0 +1,39 @@
+
+{
+    'name': 'Sale Automatic Workflow',
+    'version': '8.0.0.2.0',
```
version must be **9.0.0.0.0**

pedrobaeza replied:
Better 9.0.1.0.0 :wink:

guewen commented on sale_automatic_workflow/sale.py:
```diff
@@ -0,0 +1,88 @@
+                                          string='Automatic Workflow',
+                                          ondelete='restrict')
+
+    def _prepare_invoice(self, cr, uid, order, lines, context=None):
```
this method has been changed in 9.0:

```
    @api.multi
    def _prepare_invoice(self):
        ...
```

guewen commented on sale_automatic_workflow/sale.py:
```diff
@@ -0,0 +1,88 @@
+from openerp import models, fields, api, _
+
+
+class sale_order(models.Model):
```
Should be `SaleOrder` as it uses the new API.

guewen commented on sale_automatic_workflow/sale.py:
```diff
@@ -0,0 +1,88 @@
+            invoice_vals['journal_id'] = workflow.property_journal_id.id
+        return invoice_vals
+
+    def _prepare_order_picking(self, cr, uid, order, context=None):
```
`_prepare_order_picking` does no longer exist in 9.0.
You need something like that:

``` python
class StockMove(models.Model):
    _inherit = 'stock.move'

    @api.model
    def _prepare_picking_assign(self, move):
        """ Prepares a new picking for this move as it could not be assigned to
        another picking. This method is designed to be inherited.
        """
        values = super(StockMove, self)._prepare_picking_assign(move)
        if move.procurement_id.sale_line_id:
            sale = move.procurement_id.sale_line_id.order_id
            values['workflow_process_id'] = sale.workflow_process_id.id
        return values

```

lmignon commented:
It seems that the history of changes is lost... Not sure that we can merge this branch

```
Commits that don’t share a common merge base with the base branch are not supported.
You can view the full commit page here instead.
```

atchuthan commented:
@lmignon we don't have this module available in version 9.0, so we added the commit histories from 8.0 branch using https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-9.0

atchuthan commented:
FYI, we started working on this PR.  So, could we have "Work in progress" label?

lmignon commented:
@atchuthan OK thank you for the explanation!

guewen commented:
> FYI, we started working on this PR. So, could we have "Work in progress" label?

Done!

elicoidal commented on sale_automatic_workflow/__init__.py:
```diff
@@ -0,0 +1,29 @@
+# -*- encoding: utf-8 -*-
```
Please use simplified headers

xavier-dass committed:
[WIP] : Automated Workflows

guewen commented:
@atchuthan what's your progress? Can I help for the migration?

yvaucher commented on sale_automatic_workflow/automatic_workflow_job.py:
```diff
@@ -0,0 +1,122 @@
+        _logger.debug('Sale Orders to validate: %s', sales)
+        for sale in sales:
+            with commit(self.env.cr):
+                sale.action_button_confirm()
```
the method is now: `action_confirm`

yvaucher commented on sale_automatic_workflow/automatic_workflow_job.py:
```diff
@@ -0,0 +1,122 @@
+        _logger.debug('Invoices to validate: %s', invoices)
+        for invoice in invoices:
+            with commit(self.env.cr):
+                invoice.signal_workflow('invoice_open')
```
Workflow have been removed

atchuthan committed:
[FIX] fixed conventions

atchuthan commented:
@guewen sorry for the late reply as we were on a trip... 
We have made few changes to the code as per the discussions in the community thread.

atchuthan committed:
[FIX] modified data file

guewen commented on sale_automatic_workflow/models/account_invoice.py:
```diff
@@ -0,0 +1,24 @@
+    )
+    # TODO propose a merge to add this field by default in account module
+    sale_ids = fields.Many2many(
+        comodel_name='sale.order',
```
The counterpart field on `sale.order` is a computed field so does this `Many2many` work as expected?

`addons/sale/sale.py`:

``` python
     invoice_ids = fields.Many2many("account.invoice", string='Invoices', compute="_get_invoiced", readonly=True, copy=False)
```

guewen replied:
You probably have to go through the invoice lines to get the orders, but in fact I don't think this field is actually used.

guewen commented on sale_automatic_workflow/models/automatic_workflow_job.py:
```diff
@@ -0,0 +1,170 @@
+
+If my explanation is not clear, contact me by email and I will improve
+it: sebastien.beau@akretion.com
+"""
```
This explanation is irrelevant in 9.0, can be removed.

atchuthan committed:
[FIX] corrections and modified test

atchuthan commented:
@guewen I fixed your comments.

stephankeller commented:
Just a note that we are going to change the way the filters are defined on the automatic workflow window.  The way it is right now, somebody could delete the filter by mistake.  We are going to manage the filters the same way it is done on the "Automatic Action" window.

guewen commented on sale_automatic_workflow/models/automatic_workflow_job.py:
```diff
@@ -0,0 +1,146 @@
+    # TODO:
+    # use account_payment_mode to get the payment information
+    @api.model
+    def _register_payments(self, payment_filter):
```
Should it rather be done in sale_automatic_workflow_payment_mode https://github.com/OCA/sale-workflow/pull/297?

guewen commented on sale_automatic_workflow/models/automatic_workflow_job.py:
```diff
@@ -0,0 +1,146 @@
+                        sale_workflow.validate_invoice_filter_id.domain) +
+                    workflow_domain)
+            # TODO
+#             if sale_workflow.register_payment:
```
As above, probably to do in `sale_automatic_workflow_payment_mode`?

atchuthan replied:
yes, I agree.
I will create a PR for guewen:add-sale_automatic_workflow_payment_mode

atchuthan replied:
@guewen moved the payment functionality to sale_automatic_workflow_payment_mode and created PR to your fork https://github.com/guewen/sale-workflow/pull/3

atchuthan committed:
[FIX] modified filter usage

atchuthan committed:
[FIX] modified filter usage

atchuthan committed:
[FIX] separated payment portion

atchuthan commented:
@guewen we can change the label to "Needs review"

guewen commented:
> @guewen we can change the label to "Needs review"

yeah!

guewen commented on sale_automatic_workflow/views/sale_view.xml:
```diff
@@ -0,0 +1,28 @@
+              <field name="workflow_process_id"/>
+            </group>
+          </page>
+		</xpath>
```
Tabs spotted

guewen commented:
There is a small issue with the Configuration menu (same issue in `sale_exception` btw):

![selection_032](https://cloud.githubusercontent.com/assets/417223/15213612/a426967e-1848-11e6-9ecd-654e4911d9ae.png)

guewen commented:
I really like the feature to customize the filters, well done!

Thanks :)

2 small things to fix, but :+1:

atchuthan committed:
[FIX] corrections

atchuthan committed:
[MVD] moved payment filter to sale_automatic_workflow_payment_mode

Commit:
Fix default filters

The addon does not install if the table sale_workflow_process contains
records without filters, Odoo tries to get the xmlids of the default
values before the filters are created.

guewen commented:
@atchuthan here's a fix for a corner case during the installation (existing records in `sale.workflow.process`, `default` are called before the filters are created): https://github.com/sodexis/sale-workflow/pull/1

atchuthan committed:
Merge pull request #1 from guewen/sodexis-9.0-sale_automatic_workflow

Fix default filters

atchuthan commented:
@guewen thanks for the fix.

gurneyalex commented:
:+1:

rafaelbn commented on sale_automatic_workflow/README.rst:
```diff
@@ -0,0 +1,79 @@
+  * Validate the invoice
+  * Confirm the picking
+
+This module is used by Magentoerpconnect and Prestashoperpconnect.
```
@pedrobaeza @sergio-teruel take a look for Prestashoperpconnect

rafaelbn commented:
Nice work thanks. :+1:

rousseldenis commented:
@atchuthan I try to review the v10 migration, but I need some explanation about the filter (the char field).

Why having removed the char_domain widget ? So, as the char field is not used in code (filter_id is used), why isn't it in readonly?

atchuthan commented:
@rousseldenis sorry for the delay

first, we had a thought of using the char field directly but later due to time constraint, we worked directly with the filter.
Our idea was to use this char field where the users could setup their custom filter domain instead of making modifications to the filter domain directly.

Let us know if you need any support from our side. Thanks and let us know if you have started PR for v10 migration.

rousseldenis commented:
@atchuthan The migration to v10 has been started by @lasley 

I want to port this discussion to https://github.com/OCA/sale-workflow/pull/391
